### PR TITLE
fix(server): extraction of Samsung Motionphoto videos

### LINKED ISSUE #4873: [BUG] Unable to play motion photos from One UI 6 (Android 14)
### The bug

Immich is unable to play motion photos captured from One UI 6 (Samsung's Android 14) while all other apps can play it just fine.
I'm not sure if this applies to all Android 14 devices or if it is Samsung specific.

While I am unsure of the exact cause, the following characteristics were noticed for all photos taken with that particular OS version:

One UI 5 (Android 13) has
Directory item length : 0
Directory item padding : 59

One UI 6 (Android 14) has
Directory item length : varies from file to file
Directory item padding : 111

I added one sample picture from each version.
https://drive.google.com/drive/folders/1as4MVNJsLQWUTeVFcv8AusSukWHEFaDb?usp=sharing

Support thread from discord:
https://discord.com/channels/979116623879368755/1170085304414240848

### The OS that Immich Server is running on

Synology Docker Compose

### Version of Immich Server

v1.84

### Version of Immich Mobile App

v1.84

### Platform with the issue

- [X] Server
- [X] Web
- [X] Mobile

### Your docker-compose.yml content

```YAML
version: "3.8"

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: ["start.sh", "immich"]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - /volume1/homes/user/Photos:/data/Photos:ro
    env_file:
      - .env
    depends_on:
      - redis
      - database
      - typesense
    restart: always

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends:
    #   file: hwaccel.yml
    #   service: hwaccel
    command: ["start.sh", "microservices"]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - /volume1/homes/user/Photos:/data/Photos:ro
    env_file:
      - .env
    depends_on:
      - redis
      - database
      - typesense
    restart: always

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    volumes:
      - model-cache:/cache
    env_file:
      - .env
    restart: always

  immich-web:
    container_name: immich_web
    image: ghcr.io/immich-app/immich-web:${IMMICH_VERSION:-release}
    env_file:
      - .env
    restart: always

  typesense:
    container_name: immich_typesense
    image: typesense/typesense:0.24.1@sha256:9bcff2b829f12074426ca044b56160ca9d777a0c488303469143dd9f8259d4dd
    environment:
      - TYPESENSE_API_KEY=${TYPESENSE_API_KEY}
      - TYPESENSE_DATA_DIR=/data
      # remove this to get debug messages
      - GLOG_minloglevel=1
    volumes:
      - tsdata:/data
    restart: always

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:70a7a5b641117670beae0d80658430853896b5ef269ccf00d1827427e3263fa3
    restart: always

  database:
    container_name: immich_postgres
    image: postgres:14-alpine@sha256:28407a9961e76f2d285dc6991e8e48893503cc3836a4755bbc2d40bcc272a441
    env_file:
      - .env
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      - pgdata:/var/lib/postgresql/data
    restart: always

  immich-proxy:
    container_name: immich_proxy
    image: ghcr.io/immich-app/immich-proxy:${IMMICH_VERSION:-release}
    ports:
      - 2283:8080
    depends_on:
      - immich-server
      - immich-web
    restart: always

volumes:
  pgdata:
  model-cache:
  tsdata:
```

### Your .env content

```Shell
# You can find documentation for all the supported env variables at https://immich.app/docs/install/environment-variables

# The location where your uploaded files are stored
UPLOAD_LOCATION=./library

# The Immich version to use. You can pin this to a specific version like "v1.71.0"
IMMICH_VERSION=release

# Connection secrets for postgres and typesense. You should change these to random passwords
TYPESENSE_API_KEY=some-random-text
DB_PASSWORD=postgres

# The values below this line do not need to be changed
###################################################################################
DB_HOSTNAME=immich_postgres
DB_USERNAME=postgres
DB_DATABASE_NAME=immich

REDIS_HOSTNAME=immich_redis
```

### Reproduction steps

```bash
1. Take a picture with One UI 6 device
2. Upload it to Immich, doesn't matter web or mobile
3. While the image itself loads, attempting to play motion photo will result in infinite looping of video, and trying to hover the cursor on the motion photo icon on the timeline will show an error mark.
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by kaysond**
Fixes #4873.

The issue is that the current video extraction relies on data length and padding from the XMP inside the EXIF. This happened to work on One UI 5 (the resulting mp4 file was technically invalid, but still worked because the extra bytes were at the end), but stopped working on One UI 6. In any case, the author of exiftool recommends not using the XMP because it doesn't have any fundamental connection to the "Trailer" of the jpeg where the video data actually resides. See https://exiftool.org/forum/index.php?topic=15565.0

Fortunately, exiftool is already able to parse the jpeg trailer data structure, and exiftool-vendored has methods to leverage that. The data structure is somewhat complicated, (see https://github.com/exiftool/exiftool/blob/master/lib/Image/ExifTool/Samsung.pm#L1532), so I think it's unwise to attempt to parse the structure ourselves.

This also adds support for extracting the video from an heic-encoded motionphoto, since it's just a different tag name. (see https://exiftool.org/forum/index.php?topic=7161)

This is a minimally-invasive but naive implementation. Creating a dedicated exiftool singleton for each video that has to be extracted is inefficient and the performance hit would probably be noticeable if you upload a lot of motionphotos at once. Fixing this pretty straightforward; it would just mean refactoring exiftool up from `metadata.repository` into `metadata.service`.

I also didn't touch any of the existing extraction code. I'm not sure if it's still needed for iOS or other motionphoto implementations (Google?), but it could potentially be removed if not.

### CODE REVIEW

File: server/src/domain/repositories/metadata.repository.ts [hunk 1]
```diff
@@ -35,4 +38,5 @@ export interface IMetadataRepository {
   reverseGeocode(point: GeoPoint): Promise<ReverseGeocodeResult | null>;
   readTags(path: string): Promise<ImmichTags | null>;
   writeTags(path: string, tags: Partial<Tags>): Promise<void>;
+  extractBinaryTag(tagName: string, path: string): Promise<Buffer>;
```
Comment (jrasm91) [inline server/src/domain/repositories/metadata.repository.ts:41]: Can you have it be `path` and then `tagName`?

**Review by jrasm91 — APPROVED** [review]
Looks good, minus the linting.

**Comment by kaysond** [comment]
Done with the draft. Just need to actually test it now and it should be good to go.

Do we want to add unit tests for live/motionphoto video extraction?

File: server/src/domain/repositories/metadata.repository.ts (hunk 1 above)
Comment (kaysond) [reply server/src/domain/repositories/metadata.repository.ts:41]: Done

**Comment by jrasm91** [comment]
If you could add a unit test or two that would be great.

**Review by danieldietzler — APPROVED** [review]
LGTM! :)

**Comment by kaysond** [comment]
> LGTM! :)

Thanks. Going to make one more change so that running the metadata extraction job will fix broken videos.

File: server/src/domain/metadata/metadata.service.ts [hunk 2]
```diff
@@ -381,7 +381,11 @@ export class MetadataService {
       length = videoOffset;
     }
 
-    if (!length) {
+    if (
+      length === 0 &&
+      tags.MotionPhotoVideo === undefined &&
+      (tags.EmbeddedVideoFile === undefined || tags.EmbeddedVideoType !== 'MotionPhoto_Data')
+    ) {
```
Comment (jrasm91) [inline server/src/domain/metadata/metadata.service.ts:388]: Can you require this as: not length or no motion tag then return? Maybe it has motion tag as a constant in the line above

File: server/src/domain/metadata/metadata.service.ts [hunk 3]
```diff
@@ -423,6 +440,11 @@ export class MetadataService {
         await this.jobRepository.queue({ name: JobName.METADATA_EXTRACTION, data: { id: motionAsset.id } });
       }
 
+      if (asset.livePhotoVideoId) {
```
Comment (jrasm91) [inline server/src/domain/metadata/metadata.service.ts:443]: Two cases - corrupted or good extract

**Comment by kaysond** [comment]
Ok this is finally ready for merging. Thanks to everyone who helped!

A few notes:

- Users can run the Metadata Extraction (All) job to fix corrupt motionphoto videos. Maybe we can mention this in the release notes @alextran1502 ?
- Because we fix corrupt videos, the above job has slowed down from before. It could be moved into an entirely separate job, but that would be a much bigger change.
- It seems that the metadata extraction job does not run on trashed assets, but it does on the trashed asset's motionphoto videos. This means errors can still show up if an image with corrupt motionphoto video is trashed. (NB: it may be worth updating the "trashing" job to also trash the associated video)

**Review by jrasm91 — COMMENTED** [review]
I'm surprised this works because motionPath is the same for both the old and new motion assets. Saving the new video would overwrite the old motion file, but then you queue the old one for deletion.

**Comment by kaysond** [comment]
> I'm surprised this works because motionPath is the same for both the old and new motion assets. Saving the new video would overwrite the old motion file, but then you queue the old one for deletion.

@jrasm91 - that's a really good catch. I'm guessing that by the time the code looked like this, my immich instance had all the right assets, so it never triggered that code.

I moved the delete before the save. There's no race condition here, right? Meaning `await this.jobRepository.queue({ name: JobName.ASSET_DELETION, data: { id: asset.livePhotoVideoId } });` will block until the job completes?

**Comment by jrasm91** [comment]
No, that only queues it. Who knows when it will be deleted. Although, we can also solve this by simply changing how the original path is generated - it doesn't have to be the parent asset's id. I think we could generate a UUID ahead of time, and use that to create the motion asset and motion asset original path.

**Comment by kaysond** [comment]
> No, that only queues it. Who knows when it will be deleted. Although, we can also solve this by simply changing how the original path is generated - it doesn't have to be the parent asset's id. I think we could generate a UUID ahead of time, and use that to create the motion asset and motion asset original path.

Ok this should take care of things. I've been trying to test it but I can't get the dev docker compose to start up because of some weird docker issue. Tried two different machines and same bug on both...

If you have a chance to test it please lmk.

**Comment by kaysond** [comment]
This should finally be ready for merge @jrasm91 
Tested by:

1. Ran immich v1.93.2 dev instance
2. Uploaded 3 test motionphotos
3. Note that OneUI 5 works, but 6 is broken
4. Switch to PR dev instance
5. Note that OneUI 5 still works, but 6 is broken
6. Run metadata extraction job
7. Note that old video assets are deleted and new video assets are extracted correctly
8. Run job again, nothing happens as expected

**Comment by kaysond** [comment]
I'm going to hold my tongue until this is actually merged :D

Formatting fixed and tests fixed and updated to have more

**Review by jrasm91 — APPROVED** [review]
I tested this and it worked well. The only thing I noticed was that the broken video could be cached in the browser and that could prevent it from playing successfully immediately afterwards.

**Pull request merged**